dascore/core/spool.py

```diff
@@ -94,6 +95,12 @@ def get_contents(self) -> pd.DataFrame:
         Get a dataframe of the patches that will be returned by the spool.
         """
 
+    @abc.abstractmethod
```

Comment:
We need to think about whether we want to require all spools to implement sort. I am thinking about a spool that interfaces with a remote server (imagine if you are trying to interact with all of EarthScope's DAS data for example); maybe it doesn't make sense to require it to to know how to sorts its own massive contents. 

For now, rather than making this an abstractmethod, which will require all subclasses to implement `sort`, let's change it to just raise a `NoteImplementedError`: 

```python
def sort(self, attribute) -> Self:
    """
    Sort the Spool based on a specific attribute.
    """
    raise NoteImplementedError(f"spool of type {self.__class__} has no sort implementation")
```

Then if the subclass implements sort it will work fine, otherwise this Error will get raised.

Reply:
Good idea!

Reply:
I included this raise using a try/except block. However, I'm not sure how I can test it. For testing, I need a spool whose type has no sort implementation. Do you have any suggestions?

Reply:
I think just a dummy spool for testing will work.

Reply:
If you just want to make sure it raises for the sake of testing you can call it directly and pass an instance of a subclass spool to it. 

```python 
def test_base_spool_sort_raises(self, random_spool):
    with pytest.raises(...):
        BaseSpool.sort(random_spool, 'time')
```

Reply:
The test failed. Here is the test:
```python
    def test_base_spool_sort_raises(self, random_spool):
        expected_str = "spool of type"
        with pytest.raises(NotImplementedError, match=expected_str):
            BaseSpool.sort(random_spool, 'time')
```

Error:
```python
    def test_base_spool_sort_raises(self, random_spool):
        expected_str = "spool of type"
>       with pytest.raises(NotImplementedError, match=expected_str):
E       Failed: DID NOT RAISE <class 'NotImplementedError'>

tests/test_core/test_spool.py:243: Failed
```

Reply:
I wanted to add a test like the one below:

```python
    def test_sorting_not_implemented(self):
        """
        Test sorting a not supported spool type.
        """
        dummy_spool = ...
        expected_str = "spool of type"
        with pytest.raises(NotImplementedError, match=expected_str):
            dummy_spool.sort("time")
```

Reply:
> The test failed

Weird, what does `BaseSpool.sort` look like? Can you paste it here?

Reply:
@abc.abstractmethod
    def sort(self, attribute) -> Self:
        """
        Sort the Spool based on a specific attribute.

        Parameters
        ---------------
        atribute
            The attribute or coordinate used for sorting. If a coordinate name is used, the sorting will be based on the minimum value.
        """

Reply:
And here is the DataFrameSpool.sort:

```python
   @compose_docstring(doc=BaseSpool.sort.__doc__)
    def sort(self, attribute) -> Self:
        """
        {doc}
        """
        try:
            df = self._df
            inst_df = self._instruction_df

            # make sure a suitable attribute is entered
            attrs = set(df.columns)
            if attribute not in attrs:
                # make sure we can also cover coordinate names instead of the attribute
                if f"{attribute}_min" in attrs:
                    attribute = f"{attribute}_min"
                else:
                    msg = "Invalid attribute. Please use a valid attribute such as: 'time'"
                    raise IndexError(msg)

            # get a mapping from the old current index to the sorted ones
            sorted_df = df.sort_values(attribute).reset_index(drop=True)
            old_indices = df.index
            new_indices = np.arange(len(df))
            mapper = pd.Series(new_indices, index=old_indices)

            # swap out all the old values with new ones
            new_current_index = inst_df["current_index"].map(mapper)
            new_instruction_df = inst_df.assign(current_index=new_current_index)

            # create new spool from new dataframes
            return self.new_from_df(df=sorted_df, instruction_df=new_instruction_df)
        except IndexError:  # Catch only IndexError
            msg = "Invalid attribute. Please use a valid attribute such as: 'time'"
            raise IndexError(msg)
        except Exception:
            msg = f"spool of type {self.__class__} has no sort implementation"
            raise NotImplementedError(msg)
```

Reply:
Ok, so did you forget this line 
```python
raise NoteImplementedError(f"spool of type {self.__class__} has no sort implementation")
```
in `BaseSpool.sort`? That would explain why it isn't raising an exception in this code:

```python
def test_base_spool_sort_raises(self, random_spool):
    expected_str = "spool of type"
    with pytest.raises(NotImplementedError, match=expected_str):
        BaseSpool.sort(random_spool, 'time')
```

Reply:
> Ok, so did you forget this line

Yes. I had not added this line. It works now. Thanks very much for walking me through this!